### PR TITLE
Build on mac

### PR DESCRIPTION
**Pull request by kyessenov**
- do not set GOOS, GOARCH to incompatible values if the intent is to run the binaries natively
- do not write out of the istio repo

Signed-off-by: Kuat Yessenov <kuat@google.com>

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -37,3 +37,11 @@ install/kubernetes/istio-one-namespace.yaml
 install/kubernetes/istio.yaml
 samples/bookinfo/consul/bookinfo.sidecars.yaml
 samples/bookinfo/eureka/bookinfo.sidecars.yaml
+/bin/istio_ca
```
Comment (ldemailly) [inline .gitignore]: starting / seems odd ?

File: Makefile [hunk 2]
```diff
@@ -21,37 +21,23 @@ SHELL := /bin/bash
...
-export CGO_ENABLE=0
-
 # OUT is the directory where dist artifacts and temp files will be created.
-OUT=${GO_TOP}/out
+OUT := out
 
-GO ?= go
+# Environment for tests, the directory containing istio and deps binaries.
+# Typically same as GOPATH/bin, so tests work seemlessly with IDEs.
+ISTIO_BIN := bin
```
Comment (ldemailly) [inline Makefile]: should probably be `build/` or `output/` or such but not the same directory as an existing directory from the source tree

File: .gitignore (hunk 1 above)
Comment (kyessenov) [reply .gitignore]: that's to avoid matching nested directories

File: Makefile (hunk 2 above)
Comment (kyessenov) [reply Makefile]: it's difficult to find out all ways "bin" used throughout except by trial and error. using CI as my  linux machine.

Comment (ldemailly) [reply Makefile]: that's another good reason to use something else than bin to find sooner than later the hardcoded ones

Comment (kyessenov) [reply Makefile]: agree, but that's besides the mac build goal of this PR.

**Comment by kyessenov** [comment]
`#!/bin/bash -eo pipefail
make docker.prebuilt
cp bin/{pilot-discovery,pilot-agent,sidecar-initializer} pilot/docker
cp: cannot stat ‘bin/pilot-discovery’: No such file or directory
cp: cannot stat ‘bin/pilot-agent’: No such file or directory
cp: cannot stat ‘bin/sidecar-initializer’: No such file or directory
make: *** [docker.prebuilt] Error 1
Exited with code 2`

comments say 
`# Build docker images for pilot, mixer, ca using prebuilt binaries
docker.prebuilt:
`

@rshriram what prebuilt binaries?

**Comment by rshriram** [comment]
just run `make docker` per https://github.com/istio/istio/blob/master/DEV-GUIDE.md#building-and-pushing-the-containers .. make docker.prebuilt is only for circleci as the build job saves all the binaries...

**Comment by kyessenov** [comment]
@rshriram where's the magic that preserves the bin in circleci? I'm trying to change the location of the bin directory.

**Comment by ldemailly** [comment]
can we get this in ?
/lgtm

**Comment by rshriram** [comment]
Also as per my previous comment, docker.prebuilt is for ci. Per dev guide, you need to run make docker

**Comment by kyessenov** [comment]
Low on my priority list. Pushed to mac_build branch upstream if you want to
get it done sooner.

File: .gitignore (hunk 1 above)
Comment (costinm) [reply .gitignore]: Please don't build istio binaries in the source directory !

( we're still in process of removing the envoy binary and all the docker ones )

File: Makefile [hunk 3]
```diff
@@ -21,37 +21,23 @@ SHELL := /bin/bash
...
-export GOPATH
-
-# If GOPATH is made up of several paths, use the first one for our targets in this Makefile
-GO_TOP := $(shell echo ${GOPATH} | cut -d ':' -f1)
-
-export CGO_ENABLE=0
-
 # OUT is the directory where dist artifacts and temp files will be created.
-OUT=${GO_TOP}/out
+OUT := out
```
Comment (costinm) [inline Makefile]: This affects non-mac platforms - please don't change the way we build for other platforms, and I don't think mac needs to build in the source tree either.

File: Makefile (hunk 2 above)
Comment (costinm) [reply Makefile]: Don't know about mac build goals - but this is affecting non-mac builds, which is not good./

File: Makefile [hunk 4]
```diff
@@ -585,15 +571,15 @@ ${OUT}/istio-sidecar.deb: ${ISTIO_BIN}/envoy ${ISTIO_BIN}/pilot-agent ${ISTIO_BI
 	   --config-files /var/lib/istio/envoy/sidecar.env \
 	   --config-files /var/lib/istio/envoy/envoy.json \
 	   --description "Istio" \
-	   src/istio.io/istio/tools/deb/istio-start.sh=/usr/local/bin/istio-start.sh \
-	   src/istio.io/istio/tools/deb/istio-iptables.sh=/usr/local/bin/istio-iptables/sh \
-	   src/istio.io/istio/tools/deb/istio.service=/lib/systemd/system/istio.service \
-	   src/istio.io/istio/security/tools/deb/istio-auth-node-agent.service=/lib/systemd/system/istio-auth-node-agent.service \
+	   tools/deb/istio-start.sh=/usr/local/bin/istio-start.sh \
```
Comment (costinm) [inline Makefile]: What does this have to do with mac build ??

**Comment by kyessenov** [comment]
Where do we build binaries if not in the repo? I don't like this build
overriding my go binaries and my home locations.

File: bin/gobuild.sh [hunk 5]
```diff
@@ -59,5 +50,5 @@ while read line; do
 done < "${BUILDINFO}"
 
 # forgoing -i (incremental build) because it will be deprecated by tool chain. 
-time GOOS=${GOOS} GOARCH=${GOARCH} ${GOBINARY} build ${V} ${GOBUILDFLAGS} -o ${OUT} \
+time CGO_ENABLED=0 ${GOBINARY} build ${V} ${GOBUILDFLAGS} -o ${OUT} \
```
Comment (costinm) [inline bin/gobuild.sh]: It looks like another case where non-MAC builds will be affected. Please restrict the change to mac builds, and open separate PRs for changing how we build on linux.

**Comment by kyessenov** [comment]
Can we have a circle CI defined variable that sets the output? Clearly, we
have no option between polluting the system and polluting the repo, that
needs to be resolved with a variable.

**Comment by costinm** [comment]
Clearly we have plenty of better options - use $OUT/$OS/$ARCH for example
for mac, since
we are trying to move all artifacts and temp files under $OUT ( which is
currently $GOPATH/out ).
Docker will also eventually build there ( instead of copying binaries in
various locations in the source tree).

All you need to change is not generating linux binaries in $GOPATH/bin for
mac. No need to change
how linux is built for that, it's pretty easy to set a variable only for
mac.

**Comment by kyessenov** [comment]
What if $GOPATH set to /usr/local that does not even have write permissions?
I think it's ok to write to repo directory for local development. We can
also add a symlink like what bazel does. I can customize the makefile for
local specifically, but it is making makefile more complicated that it
needs to be.

**Comment by mattdelco** [comment]
If I were to hack things up and pretend like I cared about go devs want to use IDEs that expect bin/ then I'd try something like this:

https://github.com/istio/istio/compare/master...mattdelco:master8?expand=1

**Comment by costinm** [comment]
If GOPATH is set to /usr/local - we can warn or the build will fail. Not a
problem, the layout of
a go project and use of GOPATH is documented pretty well.

It is not generally a good practice to 'dirty' the source repository with
binaries, and for the
purpose of fixing the mac problems you don't need to do more than moving
the cross compiled
binaries ( if mac is detected ) under $OUT. The change should be ~10 lines
of makefile - which
is not that more complicated.

**Comment by kyessenov** [comment]
The goal is _not to write Linux binaries into my home_. I have to change
existing rules to stop doing that.

**Comment by costinm** [comment]
You need to override ISTIO_BIN if mac is detected. That's all - you don't
need to change rules
for this.

**Comment by kyessenov** [comment]
It turns out we hardcode "bin" and "GOPATH/bin" in a few places. Changing
just makefile is not enough.

**Comment by costinm** [comment]
Sure, if we hardcode them we should fix it.

'bin' by itself should obviously not be used.

**Comment by kyessenov** [comment]
closing, please re-open using mac_build branch

**Comment by ldemailly** [comment]
:-(
